### PR TITLE
Initialize ModelCheckpoint state as early as possible

### LINKED ISSUE #11110: ModelCheckpoint state does not get reloaded after moving `on_init_end` implementation
## 🐛 Bug

Discovered in #10995. Present only in master. Appeared after #10940.

The state in ModelCheckpoint does not get reloaded because the state_key does not match. This is a regression caused by #10940 which has moved the code in `ModelCheckpoint.on_init_end` to `ModelCheckpoint.on_pretrain_routine_start`, causing the Trainer to attempt reloading the state BEFORE the state_key is fully defined. In particular, the `save_on_train_epoch_end` key gets determined lazily.

The bug is not present in any release.

### To Reproduce

This can be reproduced by setting  `max_epochs=1` in stead of `max_steps=1` in `tests/models/test_restore.py::test_callbacks_state_fit_ckpt_path`.

### Expected behavior

Callback state gets reloaded when calling Trainer.fit and ModelCheckpoint arguments are unmodified.

### Environment

You can also fill out the list below manually.
-->

- PyTorch Lightning Version (e.g., 1.5.0): master (15.12.2021)
- PyTorch Version (e.g., 1.10): 1.10
- Python version (e.g., 3.9):
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

cc @tchaton @carmocca @awaelchli @ninginthecloud @jjenniferdai @ananthsub @daniellepintz @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #11110
Fixes a combination of two bugs discovered in #10995. 

1. The test `test_callbacks_state_fit_ckpt_path` ran with max_steps=1 instead of `max_epochs=1` never creating a checkpoint. This had the effect of hiding bug 2:
2. The state does not get reloaded because the state_key does not match. This is a regression caused by #10940 which has moved the code in `ModelCheckpoint.on_init_end` to `ModelCheckpoint.setup`, causing the Trainer to attempt reloading the state BEFORE the state_key is fully defined. In particular, the `save_on_train_epoch_end` key gets determined lazily.

This PR fixes these two issues by setting max_epochs=1 in the test and moving the implementation of `ModelCheckpoint.on_pretrain_routine_start` to `ModelCheckpoint.on_before_accelerator_backend_setup` which runs BEFORE callback states are reloaded. 

**The latter is a hotfix, but a discussion should be held whether we want to reintroduce/de-deprecate on_init_end.**
@ananthsub @daniellepintz @justusschock 

The bug is not present in any release. Milestone 1.6.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @borda @tchaton @carmocca @awaelchli @ninginthecloud @jjenniferdai

### CODE REVIEW

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -248,13 +248,16 @@ def state_key(self) -> str:
             save_on_train_epoch_end=self._save_on_train_epoch_end,
         )
 
-    def on_pretrain_routine_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
-        """When pretrain routine starts we resolve the ckpt dir on the fly."""
+    def on_before_accelerator_backend_setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_checkpoint.py:251]: how about running this in the `setup` hook?

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:251]: yes agreed better than the accelerator_setup hook

**Review by ananthsub — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by daniellepintz** [comment]
super nit from the description:
> which has moved the code in ModelCheckpoint.on_init_end to ModelCheckpoint.setup

do you mean "which has moved the code in ModelCheckpoint.on_init_end to ModelCheckpoint.on_pretrain_routine_start"?

**Comment by awaelchli** [comment]
Yes, typo in update. Was meant to say that this PR moves `ModelCheckpoint.on_pretrain_routine_start` to `ModelCheckpoint.setup`

**Pull request merged**